samples/panther/ap/src/mqtt.c

```diff
@@ -0,0 +1,389 @@
+/*
+ * Copyright (c) 2017 IpTronix S.r.l.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <net/mqtt.h>
+
+#include <net/net_context.h>
+
+#include <misc/printk.h>
+#include <string.h>
+#include <errno.h>
+
+#include "mqtt.h"
+
+#define ZEPHYR_ADDR       "0.0.0.0"
+
+#define APP_SLEEP_MSECS	  5000
+#define APP_TX_RX_TIMEOUT  800
+#define APP_CONNECT_TRIES	10
+
+/**
+ * @brief mqtt_client_ctx	Container of some structures used by the
+ *				publisher app.
+ */
+struct mqtt_client_ctx {
+	/**
+	 * The connect message structure is only used during the connect
+	 * stage. Developers must set some msg properties before calling the
+	 * mqtt_tx_connect routine. See below.
+	 */
+	struct mqtt_connect_msg connect_msg;
+	/**
+	 * This is the message that will be received by the server
+	 * (MQTT broker).
+	 */
+	struct mqtt_publish_msg pub_msg;
+
+	/**
+	 * This is the MQTT application context variable.
+	 */
+	struct mqtt_ctx mqtt_ctx;
+
+	/**
+	 * This variable will be passed to the connect callback, declared inside
+	 * the mqtt context struct. If not used, it could be set to NULL.
+	 */
+	void *connect_data;
+
+	/**
+	 * This variable will be passed to the disconnect callback, declared
+	 * inside the mqtt context struct. If not used, it could be set to NULL.
+	 */
+	void *disconnect_data;
+
+	/**
+	 * This variable will be passed to the publish_tx callback, declared
+	 * inside the mqtt context struct. If not used, it could be set to NULL.
+	 */
+	void *publish_data;
+};
+
+/* This is the network context structure. */
+static struct net_context *net_ctx;
+
+/* The mqtt client struct */
+static struct mqtt_client_ctx client_ctx;
+
+/* Status of mqtt */
+static int mqtt_status = 0;
+
+/* This routine sets some basic properties for the network context variable */
+static int network_setup(struct net_context **net_ctx, const char *local_addr,
+			 const char *server_addr, u16_t server_port);
+
+/* The signature of this routine must match the connect callback declared at
+ * the mqtt.h header.
+ */
+static void connect_cb(struct mqtt_ctx *mqtt_ctx)
+{
+	struct mqtt_client_ctx *client_ctx;
+
+	client_ctx = CONTAINER_OF(mqtt_ctx, struct mqtt_client_ctx, mqtt_ctx);
+
+	printk("[%s:%d]", __func__, __LINE__);
```

Comment:
I'd rather use the logging stuff for debugging messages.  (There are other places in the code where printk() is used and the logging subsystem could be used instead.)